docs/internal/interop.md

```diff
@@ -0,0 +1,351 @@
+the sandbox code can use for everything. The sandbox code can also access that
+memory and it uses to keep the arguments and return values, so that the sandbox
+can see them.
+
+Because that heap is in a Verona sandbox region, it is safe to assume there are
+no race conditions introduced by the compiler onto the external library, due to
+the use of cowns to execute the foreign code.
+
+## Step-by-step example
+
+This is the description of a mock implementation of the child process sandbox
+running the C++ snippet in the beginning of this document. We'll go line by line
+and describe what the compiler and the generated code will do for each of them.
+
+### Compile time
+
+```c
+type StdVec = "c++::vector";
+```
+
+* Call the C++ driver (clang wrapper) to parse the include file (and all its
+  includes) and keep as a C++ AST.
+* Create a Verona class called `StdVec` with C++ built-in types defined as Verona
+  types (ex. `int` as `S32`, `size_t` as `U64`).
+* Expose an interface for the parser to query symbols (types, objects and
+  functions).
+
+```c
+reg = new StdVec;
+```
+
+* The parser recognises the region as a C++ sandbox.
+* Lower code to create a region with a sandbox allocator (RT calls).
+* Lower code to create the sandbox memory area (RT allocator calls).
+* Create a wrapper object for function dispatch and general sandbox utilities.
+* The region is bound to the variable `reg`.
+
+```c
+obj = new std::vector[int] in reg
+```
+
+* The parser recognises the region and tries to create a new object in the
+  sandbox region by using the C++ driver.
+* Assemble the foreign call signature (namespace, template args):
+  `std::vector<int> std::vector<int>::vector<int>()`
+* Instantiate the signature as a template specialisation in the AST, using
+  the query interface to validate the code and get an actual implementation.
+* If available, create the Verona equivalent inside the `StdVec` class and
+  Lower call to that function. Otherwise emit an error and stop processing.
+* The Verona signature would be something like
+  `create() : StdVec::std::vector[S32]`.
+* Generate the wrapper function from memory buffer to return value.
+* Add that function to the sandbox dispatcher's table and associate the index
+  with this particular specialisation, increment the index.
+* On the recently created `StdVec` Verona function, lower Verona code to call the
+  dispatcher with the index and the buffer and to extract the return value.
+* Bind the result of the Verona call to the variable `obj`.
+
+
+```c
+obj.push_back(42)
+```
+
+* The parser recognises the object is in a sanbdox region.
+* Assemble the foreign call signature (namespace, template args):
+  `void std::vector<int>::push_back(int)`
+* Same as above, creates a new Verona function in the `StdVec` class that will
+  serialise arguments and return value and call the dispatcher.
+* The Verona signature would be something like
+  `StdVec::std::vector[S32]::push_back(S32)`.
```

Comment:
As above, we don't want to use `S32` because that binds the Verona code to a specific sandbox ABI, which has all sorts of problems.

Reply:
Each signature would be specific to the program/foreign language/hardware architecture. The same program and the same library compiled on a different platform could have a completely different type.

Calling it `ProcessSandbox::int` or using `S32`, for a specific instance of a program, is irrelevant if they both have the same sign, size, alignment and semantics.

It could be misleading having a single `ProcessSandbox::int` type name with multiple sizes, just as it is in C++, but that will have to adapt to multiple different foreign languages.

Reply:
`S32` is the same size on all platforms.  `ProcessSandbox::int` requires an explicit cast to and from *any* Verona native type.  This is necessary to be able to write portable the Verona code that handles these (I've made this mistake before, I don't want to do it again).

Reply:
Changed both to `Sandbox::int` and `Sandbox::size_t`.

Reply:
I think you're misunderstanding my point. To be explicit, I didn't mean to force the size here at a source level. This would be madness.

I meant in a particular platform (OS, arch, ABI) the size is `S32` so a binary of that platform, `Sandbox::int` will end up being with an identical representation as `S32`, so my example had that to mean what the actual size would be in this fictitious platform. This is only referring to this particular "already compiled" representation, and not the source code.